docs/experimental-features.md

```diff
@@ -0,0 +1,286 @@
+
+- The feature stages definition
+- The process from the inception to the graduation of the feature to stable
+- Some technical guidelines on how to include the feature, including how to add
+  new API fields and how to turn on/off experimental features.
+
+Some technical details are left unspecified on purpose, because they tend to
+depend on the features themselves.
+
+## Scope
+
+Experimental features might consist in:
+
+- Behavioural/Semantic additions/changes, without affecting the APIs (CRDs).
+  E.g. [#4560](https://github.com/knative/eventing/pull/4560)
+- Additive changes to the APIs, adding new behaviours. E.g.
+  [#5148](https://github.com/knative/eventing/issues/5148)
+
+This process MUST be followed in all the situations above explained for changes
+affecting the https://github.com/knative/eventing repository. It's heavily
+recommended, but not required, for all the Eventing subprojects to follow this
+process too.
+
+Breaking changes to the APIs, like modifying the API object structure, or
+complete new CRDs, are _not covered_ by this document and should be introduced
+using the usual API versioning, e.g. bumping from _v1_ to _v2alpha1_ or adding a
+new CRD as _v1alpha1_.
+
+## Runtime feature gate
+
+We store in a `ConfigMap` a map of enabled/disabled features. These are very
+similar to
+[Kubernetes Feature Gates](https://kubernetes.io/docs/reference/command-line-tools-reference/feature-gates/)
+and
+[Knative Serving feature flags](https://github.com/knative/serving/blob/master/pkg/apis/config/features.go).
+When the feature is implemented, we provide a Golang API to check if the feature
+is enabled or not, like `experimental.FromContext(ctx).IsEnabled(featureName)`.
+
+The user can easily enable the features modifying a config map/environment
+variable of her/his knative setup.
+
+<!-- TODO add more details when the pr is merged-->
+
+## Strategies to add new APIs for experimental features
+
+There are three main **recommended** strategies to implement APIs for new
+features:
+
+- If the feature requires a field on the API, or the feature may affect several
+  parts of the API, modify the stable APIs adding the necessary fields. These
+  fields **must** be optional, and it should be clearly stated in the godoc that
+  the feature is experimental and may change. In the CRD yaml the field(s)
+  should not be specified, and `x-kubernetes-preserve-unknown-fields` should be
+  used instead. Then, in the webhook, you can reject resources with fields
+  related to experimental features when validating the input CR using
+  [experimental.ValidateAPIFields()](https://github.com/knative/eventing/blob/917f57f0e68abbd0256c5fa4bff1ce2e83ffe78b/pkg/apis/experimental/api_validation.go#L13)
+- Alternative to the above approach, If the feature affects a whole Resource,
+  and the API is a single value, that is not an object or an array of values,
+  use an annotation. Then, in the webhook, you can reject resources with
```

Comment:
IMO, the choice between annotations vs spec field(s) depends on whether the feature is normative or not. An annotation might graduate into a field if it gains general consensus and applicability.

Reply:
See comment below